[URGENT] Fix sanitizer build

levitte opened the pull request:
## DECODER EVP_PKEY: Don't store all the EVP_KEYMGMTs

OSSL_DECODER_CTX_new_by_EVP_PKEY() would keep copies of all the
EVP_KEYMGMTs it finds.
This turns out to be fragile in certain circumstances, so we switch to
fetch the appropriate EVP_KEYMGMT when it's time to construct an
EVP_PKEY from the decoded data instead.  This has the added benefit
that we now actually use the property query string that was given by
the caller for these fetches.

Fixes #13503

##  MSBLOB & PVK: Make it possible to write EVP_PKEYs with provided internal key

So far, the MSBLOB and PVK writers could only handle EVP_PKEYs with
legacy internal keys.

Specially to be able to compile the loader_attic engine, we use the C
macro OPENSSL_NO_PROVIDER_CODE to avoid building the provider specific
things when we don't need them.  The alternative is to suck half of
crypto/evp/ into loader_attic, and that's just not feasible.

Fixes #13503

##  DECODER: Adjust the library context of keys in our decoders

Because decoders are coupled with keymgmts from the same provider,
ours need to produce provider side keys the same way.  Since our
keymgmts create key data with the provider library context, so must
our decoders.

We solve with functions to adjust the library context of decoded keys,
and use them.

## CORE: Separate OSSL_PROVIDER activation from OSSL_PROVIDER reference

This introduces a separate activation counter, and the function
ossl_provider_deactivate() for provider deactivation.

Something to be noted is that if the reference count goes down to
zero, we don't care if the activation count is non-zero (i.e. someone
forgot to call ossl_provider_deactivate()).  Since there are no more
references to the provider, it doesn't matter.
The important thing is that deactivation doesn't remove the provider
as long as there are references to it, for example because there are
live methods associated with that provider, but still makes the
provider unavailable to create new methods from.

Fixes #13503
Fixes #12157

##  EVP: Fix memory leak in EVP_PKEY_CTX_dup()

In most error cases, EVP_PKEY_CTX_dup() would only free the EVP_PKEY_CTX
without freeing the duplicated contents.

Fixes #13503

levitte commented:
I'm getting issues with the tests on my laptop, like this:

```
        # INFO:  @ ../master/test/evp_test.c:3166
        # ../../../master/test/recipes/30-test_evp_data/evppkey_dsa.txt:269: Expected SIGNATURE_MISMATCH got DIGESTSIGNFINAL_ERROR
        # 804BC250547F0000:error:0300009C:digital envelope routines:inner_evp_generic_fetch:unsupported algorithm:../master/crypto/evp/evp_fetch.c:345:Global default library context, Algorithm (SHA512 : 150), Properties (<null>)
        # 804BC250547F0000:error:01800078:bignum routines:BN_generate_dsa_nonce:no suitable digest:../master/crypto/bn/bn_rand.c:279:
        # 804BC250547F0000:error:05080003:dsa routines:dsa_sign_setup:BN lib:../master/crypto/dsa/dsa_ossl.c:310:
        # 804BC250547F0000:error:05080003:dsa routines:dsa_do_sign_int:BN lib:../master/crypto/dsa/dsa_ossl.c:177:
        ...
        not ok 1 - ../../../master/test/recipes/30-test_evp_data/evppkey_dsa.txt
    not ok 1 - run_file_tests
../../util/wrap.pl ../../test/evp_test -config ../../../master/test/default-and-legacy.cnf ../../../master/test/recipes/30-test_evp_data/evppkey_dsa.txt => 1
not ok 21 - running evp_test -config ../../../master/test/default-and-legacy.cnf evppkey_dsa.txt
#   Failed test 'running evp_test -config ../../../master/test/default-and-legacy.cnf evppkey_dsa.txt'
#   at ../master/test/recipes/30-test_evp.t line 130.
```

This makes me think of the recent issues with seed stuff, but I thought @paulidale had fixed it and the fix was merged.  I have *no* idead what's going on...

levitte commented:
> I have _no_ idea what's going on...

I got a better idea now, and it has nothing to do with the seed stuff.

Our decoders must adjust the library context of the created keys to match what would come out of the KEYMGMT if it was generated.

levitte commented:
Sanitizers still not happy...  that looks like a leak elsewhere, though.

levitte commented:
This is becoming a generic fixup PR...  I'll change title and description accordingly

t8m commented:
> Sanitizers still not happy... that looks like a leak elsewhere, though.

But the leak or at least its appearance seems to be triggered by this PR.

levitte commented:
> But the leak or at least its appearance seems to be triggered by this PR.

Yeah, something's up, I agree.

levitte commented:
Found 1 leak, I hope that's it.

levitte commented:
I did a sanitizer run on my laptop, using the exact same setup (apart from `no-asm`) as the sanitizer build, and it passed.  I do see that it didn't quite succeed on Github CI, there's a failure in `test/evp_extra_test2.c`, which I can't reproduce for the moment.

I'd be grateful if anyone could take a look at that last bit and help me figure it out

mattcaswell commented:
> I'd be grateful if anyone could take a look at that last bit and help me figure it out

I can periodically reproduce it with OPENSSL_TEST_RAND_ORDER=0

mattcaswell commented:
With this patch I can reliably reproduce this failure even in a build without sanitizers:

````
diff --git a/test/evp_extra_test2.c b/test/evp_extra_test2.c
index 9181061247..a78d5059fe 100644
--- a/test/evp_extra_test2.c
+++ b/test/evp_extra_test2.c
@@ -278,9 +278,9 @@ int setup_tests(void)
         return 0;
     }
 
+    ADD_TEST(test_d2i_PrivateKey_ex);
     ADD_TEST(test_alternative_default);
     ADD_ALL_TESTS(test_d2i_AutoPrivateKey_ex, OSSL_NELEM(keydata));
-    ADD_TEST(test_d2i_PrivateKey_ex);
 
     return 1;
 }
````
It looks to me like an instance of #12157

levitte commented:
Thanks for the hint, now I can go in and see what can be done about it.

paulidale commented:
There is no need for trial and error.

When running with `OPENSSL_TEST_RAND_ORDER=0`, a test failure will print "random seed = nnnnnnn".
Rerunning with `OPENSSL_TEST_RAND_ORDER=nnnnnnn` will execute the tests in the exact same order.
This should remain true across platforms.

levitte commented:
A new commit was added, which fixes #12157, and hopefully make the sanitizers happy.

("It works on my laptop!" -- famous last words)

levitte commented:
This is urgent, so it would be nice to get a review soon.

t8m commented:
It seems the old failures in the sanitizer are now fixed. There is a new one - leak in the provided test. However this is not a regression caused by this PR as the new failure is there on master already. I did not yet find out which merged PR caused it.

levitte commented:
> It seems the old failures in the sanitizer are now fixed. There is a new one - leak in the provided test. However this is not a regression caused by this PR as the new failure is there on master already. I did not yet find out which merged PR caused it.

I found that one and added a commit that I hope fixes the issue.

levitte commented:
*Now* it works on my laptop.

levitte commented:
Yay, sanitizer CI passes :smile:

paulidale commented on crypto/dh/dh_lib.c:
```diff
@@ -168,6 +168,11 @@ int DH_up_ref(DH *r)
     return ((i > 1) ? 1 : 0);
 }
 
+void ossl_dh_set0_libctx(DH *d, OSSL_LIB_CTX *libctx)
```
I'm wondering if these should also take a propq argument?  More for consistency than necessity.  I can't think why it would be needed but not having it if it is is bad.

levitte replied:
Another PR, then, and should be separate function (a set1, since we most probably will strdup it).

t8m replied:
I do not really see why we would need it. propq is attribute of an operation (or fetch for an operation) not of a key object. The pairing of libctx and propq is just a coincidence and not a real thing. We already discussed this a few months ago.

t8m commented:
Working on review but it will take me some time as the changes are quite non-trivial.

levitte commented:
> Working on review but it will take me some time as the changes are quite non-trivial.

I'd recommend reading them one commit at a time, that will at least make each change more concise.

levitte commented:
> I'd recommend reading them one commit at a time, that will at least make each change more concise.

With that in mind, should I perform a squash?

t8m commented:
> > I'd recommend reading them one commit at a time, that will at least make each change more concise.
> 
> With that in mind, should I perform a squash?

Yes, please.

levitte commented:
> > With that in mind, should I perform a squash?
> 
> Yes, please.

Done

t8m reviewed (changes requested):
Just some doc nits, otherwise LGTM.

levitte commented on doc/internal/man3/ossl_provider_new.pod:
```diff
@@ -155,9 +158,9 @@ Only text parameters can be given, and it's up to the provider to
-is:
+provider object I<prov>, by incrementing its activation count, flags
+it as activated, and initialized it if it isn't already initialized.
+Initializing means on of the following:
```
Fixed

levitte commented on doc/internal/man3/ossl_provider_new.pod:
```diff
@@ -155,9 +158,9 @@ Only text parameters can be given, and it's up to the provider to
-provider object I<prov>.
-What "activates" means depends on what type of provider object it
-is:
+provider object I<prov>, by incrementing its activation count, flags
```
Considering "incrementing", I went for "flagging" and "initializing".  I believe that makes for more consistent English (and someone that knows grammar terms better than I would probably explain it in more technical terms)

levitte replied:
Done

levitte commented on doc/internal/man3/ossl_provider_new.pod:
```diff
@@ -175,6 +178,10 @@ be located in that module, and called.
 =back
 
+ossl_provider_deactivate() "deactivates" the provider for the given
+provider object I<prov>, by decrementing its activation count.  When
```
Done

t8m commented:
Approved. Setting ready-to-merge as this is urgent.

levitte commented:
> Setting ready-to-merge as this is urgent.

Thanks.  Will merge now

levitte commented:
Merged

054cde175664f3e7c8fe5f753c0a5cb5be75dccc DECODER EVP_PKEY: Don't store all the EVP_KEYMGMTs
e77c13f8b73ff937819d6551ddd616fe01b989d0 MSBLOB & PVK: Make it possible to write EVP_PKEYs with provided internal key
6963979f5c0f95b2152ef74645faa7344e33284d DECODER: Adjust the library context of keys in our decoders
390f9bad69ce19f601abf131ceabf90aedc0d3d5 CORE: Separate OSSL_PROVIDER activation from OSSL_PROVIDER reference
74cd923a78b490d46af9c3dc5c8dc7a741c5e576 EVP: Fix memory leak in EVP_PKEY_CTX_dup()

levitte commented:
Thank you, @t8m!
